sampling/hierarchical/lib/opentelemetry/sampling/hierarchical/sampler.rb

```diff
@@ -0,0 +1,37 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Sampling
+    module Hierarchical
```

Comment:
Is Hierarchical Sampler a term used in X-Ray to describe this pattern? As opposed to a `Any` or `All` sampler?

Reply:
It is not but I find `Any` sampler to be a bit misleading since it doesn't imply a fixed order.

Reply:
I guess this name is not intuitive to me. Would you be able to help me understand what makes it Hierarchical or Ranked ordering vs "If any of the items in this list return true regardless of order"?

Reply:
The list of samplers is iterated left to right. If any returns a positive response, we stop and say that the span should be sampled. `Any` doesn't give you this guarantee (in my interpretation).
The order can make a difference. A contrived example would be having two samplers sampling the following spans
* Sampler `A`: one per second, only if `kind == :internal`
* Sampler `B`: one per second, only if `kind in [:server, :internal]`

Let's say the application produces two traces in the same second, one with `kind = :internal` and one with `kind = :server` (in this order). If sampler `A` was applied first, both traces would be sampled. On the other hand, if sampler `B` was first, only the first (`:internal`) trace would be sampled since `B` already spent all its credits when the second span arrives.